Add a note about uninhabited-struct layout optimization (#346)
rust-lang/lang-team: src/frequently-requested-changes.md

* https://github.com/rust-lang/rfcs/issues/1397
* https://github.com/rust-lang/rust/issues/17027
* https://github.com/rust-lang/unsafe-code-guidelines/issues/176

> ## Uninhabited `struct`s should all be ZSTs

Reviewer: I think the name “ZST” does not capture the underlying idea very well. The real thing we want is a -∞ sized type, ZST is just one specific way of implementing it. Although I don’t have a better title yet.

Also, do you think it’s a good idea to also add [the discussion of `Inhabitted` trait](https://github.com/rust-lang/rust/issues/93032#issuecomment-1073118304) here? In case someone reading this come up with that idea again.

Reviewer 2: > The real thing we want is a -∞ sized type

That's not as clear as you'd think. Rust has unsafe, and types need a layout, so "a ZST with always-false validity invariant" actually has some important benefits. With a -∞ sized/aligned type you can't even start executing a function that has a `!` variable in it because its whole stack frame would be uninhabited; yet it's easy to write such a function: just `panic!()`.

Reviewer 3: Size is a natural number, so a size of  -∞ doesn't even make sense. (One could define notions of size where that does make sense, but that's not the discussion we are having here. The notion of size here is the notion currently used in Rust. Given the proposed resolution for this question, it's also not useful to consider these other notions of size.)

Therefore, ZST is exactly the right term here. There's no way a type can be smaller than that.

Reviewer 4: Currently size is (at the very least) the amount of bytes required to uniquely represent an arbitrary member of the set of a type, in the Information Theory sense.

The minimum value possible there is zero: if person A needs to describe an element x of type T to person B, and in order to do so, no bytes need to be sent to B, that's because B already knows all it needs to know about x.

And that can only happen when T is either a singleton set or the empty set.

I'm guessing that the -oo notion comes from the logarithm in the Information Theory definition. If we take the logarithm of the size of the set described by the type T, we get:
- 0 if `|T| = 1`
- -oo if `|T| = 0`

But another restriction we impose over sizes is that they be positive. This (I believe) is because we need sizes to describe a layout budget.

Thus, on one hand the logarithm tells us the number of bits is at least -oo, but on the other hand we know it to also be at least 0. So it ends up being 0.

A ZST :)

Reviewer 5: Those two concerns get put together by the use of the ceiling operator; the logarithm of an integer is not necessarily an integer, even when it's not -∞, so we actually do ⌈log₂|T|⌉ where we choose to define ⌈-∞⌉=0.

> ## Uninhabited `struct`s should all be ZSTs

Reviewer 6: Just from reading this initially it's not immediately clear what ZSTs are. Maybe make this from

 > Uninhabited `struct`s should all be ZSTs

to 

> Uninhabited `struct`s should all be zero-sized types (ZSTs)

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/lang-team/346/98a2a1b432f219bf49e7ab07b68f5ab6657d4771..0c9d7a84ea1cf20d778d6c205ec5d97214023e66)*

Reviewer 7: Done.

ignoring that part here because leaking is still *sound*.)

For something simple like `make_pair::<u8, i32>`, it's clearly fine.  But with `make_pair::<u32, !>`
it's *only* sound if we *don't* let `(u32, !)` become a ZST.  We need the allocation for the box
to be large enough to write that `u32` without being an obviously-UB out-of-bounds write.

Thus if we wanted to always have uninhabited product types be ZSTs, we'd need to give up on certain
other rules, perhaps the one that `T` and `MaybeUninit<T>` always have the same size.  So far, the
simpler, less-error-prone experience for writing unsafe code has won out over the minimal space
savings possible from shrinking the types.  After all, while it's not necessarily fully unreachable,
as something like `make_pair(|| a, || loop { … })` would still need to allocate the space despite
that never reaching the `assume_init` part, it's still unlikely that this occurs frequently.

There *is* still interest in maybe doing optimizations like this on *sum* types, however.  There's more
to potentially be gained there since one variant of an `enum` being uninhabited doesn't

> keep the whole *value* from being uninhabited the way an uninhabited field does in a `struct`.

Reviewer 8: I think this might be a typo?

```suggestion
keep the whole *value* from being inhabited, the way an uninhabited field does in a `struct`.
```

Reviewer 7: Simplifying the negatives further:
```suggestion
make the whole `enum` uninhabited the way an uninhabited field does in a `struct`.
```
